[Jennifer] iP

jennibearduit opened the pull request:
# Morgan

> "Let our advance worrying become advance thinking and planning." -Winston Churchill

Morgan is a virtual task assistant that provides both **speed and convenience** in managing your day-to-day tasks.
Equipped with a simple Graphical User Interface (GUI), Morgan is optimized for use via a Command Line Interface (CLI).

Morgan frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- <s>FAST</s> SUPER FAST to use

Download Morgan [here](https://github.com/jennibearduit/ip/releases) and you'll never have to worry about missing a single deadline! 👍 

## Features
- [X] Add task
- [X] Delete task
- [X] Find tasks
- [X] List tasks

## Commands

1. `help` : lists all available commands in Morgan.


2. `list` : lists all existing tasks in storage.


3. `fixed [task] /for [duration (in hours)]` : adds a fixed duration task into the list of tasks.


4. `todo [task]` : adds a to-do task into the list of tasks.


5. `event [task] /at [dd-mm-yyyy hh:mm]` : adds an event task into the list of tasks.


6. `deadline [task] /by [dd-mm-yyyy hh:mm]` : adds a deadline task into the list of tasks.


7. `delete [task number]` : deletes a task from the list of tasks.


8. `done [task number]` marks a specified task as done.


9. `find [keyword]` : lists all existing tasks containing a specified keyword.


## Sample Code

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Implement greeting and echo

Commit:
Add task, command, parser and task manager class

Commit:
Add mark done command

Commit:
Added deadline, event and todo task

Commit:
Change input, expected and runtest

jennibearduit committed:
Add exception and commands

jennibearduit committed:
Add delete command

jennibearduit committed:
Add exceptions and commands & tasks packages

jennibearduit committed:
Edit input script

jennibearduit committed:
Add File Manager

jennibearduit committed:
Add date/time format for deadline and event task

jennibearduit committed:
Merge branch 'branch-level-7'

Finish level 7 to save task list whenever it changes.

jennibearduit committed:
Merge branch 'branch-level-8'

Finish level 8 to format date/time for deadline and event task.

jennibearduit committed:
Rename classes into Storage.java and TaskList.java and move them

jennibearduit committed:
Refactor add deadline and event commands

jennibearduit committed:
Add Ui.java

jennibearduit committed:
Fix date input format error in event and deadline tasks

jennibearduit committed:
Remove unused import

jennibearduit committed:
Remove bug and add Parser Test

jennibearduit committed:
Fix level 7 and remove bugs

jennibearduit committed:
Add JavaDoc

jennibearduit committed:
Edit to fulfill coding standard

jennibearduit committed:
Add find command

jennibearduit committed:
Merge branch 'branch-level-9'

jennibearduit committed:
Merge branch 'branch-A-CodingStandard'

jennibearduit committed:
Merge branch 'branch-A-JavaDoc'

jennibearduit committed:
Fix indentation and add JavaDoc

jennibearduit committed:
Ignore output files

jennibearduit committed:
Update imports

jennibearduit committed:
Update storage and parser

jennibearduit committed:
Rename classes and edit storage

jennibearduit committed:
Edit error messages

jennibearduit committed:
Edit task error messages

jennibearduit committed:
Update parsers

jennibearduit committed:
Change app name from Duke to Morgan

jennibearduit committed:
Merge branch 'branch-NewAppName'

jennibearduit committed:
Update .gitignore

jennibearduit committed:
Refactor Morgan.java and commands

jennibearduit committed:
Minor edit

jennibearduit committed:
Merge branch 'add-gradle-support' of github.com:jennibearduit/ip into A-Gradle

jennibearduit committed:
Merge branch 'A-Gradle'

jennibearduit committed:
Add initial gui

angnobel commented on src/main/java/Morgan.java:
```diff
@@ -0,0 +1,55 @@
+
+            // Read user input and check whether input is exit keyword
+            String userInput = ui.getInput();
+            boolean isExitKeyword = userInput.trim().equalsIgnoreCase(EXIT_KEYWORD);
```
Perhaps the trim().equalsIgnoreCase could be abstracted within parser?

nobelang78 replied:
Perhaps the trim().equalsIgnoreCase could be abstracted within parser?

angnobel commented on src/main/java/commands/FindCommand.java:
```diff
@@ -0,0 +1,51 @@
+    private static final String INPUT_FORMAT = String.format("\t\"%s [keyword]\"", KEYWORD);
+    private static final String INPUT_FORMAT_ERROR_MESSAGE = String.format("Please " +
+            "ensure your input is in the following format:\n" + INPUT_FORMAT);
+    private final String NOT_FOUND_ERROR_MESSAGE = "No matching task found. "
```
Perhaps these messages can be handled by Ui class?

nobelang78 commented on src/main/java/commands/FindCommand.java:
```diff
@@ -0,0 +1,51 @@
+    private static final String INPUT_FORMAT_ERROR_MESSAGE = String.format("Please " +
+            "ensure your input is in the following format:\n" + INPUT_FORMAT);
+    private final String NOT_FOUND_ERROR_MESSAGE = "No matching task found. "
+            + "Please try another keyword.";
```
Perhaps these messages can be handled by Ui class?

codecodecoder78 reviewed (commented):
LGTM, small abstractions that can make the code cleaner

jennibearduit committed:
Update gui

jennibearduit committed:
Add photos for user and Morgan

jennibearduit committed:
Edit logo

jennibearduit committed:
Update storage and commands

jennibearduit committed:
Fix storage bug

jennibearduit committed:
Merge branch 'Level-10'

jennibearduit committed:
Fix storage data loading

jennibearduit committed:
Refactor code

jennibearduit committed:
Refactor some code

jennibearduit committed:
Edit color

jennibearduit committed:
Update checkstyle

jennibearduit committed:
Merge branch 'A-CheckStyle'

jennibearduit committed:
Remove MainWindow.java

jennibearduit committed:
Update README.md

jennibearduit committed:
Edit README.md

jennibearduit committed:
Rename all taskList to tasks

jennibearduit committed:
Edit typo in README.md

mrmrinal commented on src/main/java/commands/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,52 @@
+ * The execution of this command will add a deadline task to the task list.
+ */
+public class AddDeadlineCommand extends AddCommand {
+    public static final String KEYWORD = "deadline";
```
I like how you used Caps letters for final constants in your class. Perhaps you could you could consider spacing out your constants by the type of operations that they do?

mrmrinal commented on src/main/java/Morgan.java:
```diff
@@ -0,0 +1,58 @@
+        this.taskList = new TaskList();
+    }
+
+    public boolean isExit(String keyword) {
```
Perhaps you can consider changing your variable name to a verb like ```hasExited ``` to make it more intuitive to the reader?

mrmrinal commented on src/main/java/Morgan.java:
```diff
@@ -0,0 +1,58 @@
+        String output;
+        Command command;
+        CommandParser cp = new CommandParser();
+        try {
```
I like how you used your try catch block in this method. Good job!

mrmrinal commented on src/main/java/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,66 @@
+        String intString = userInput.substring(KEYWORD.length()).trim();
+
+        // Checks if user specified task number
+        if (intString.isEmpty()) {
```
Perhaps you could consider throwing this statement inside the try block, and have a catch block for the MorganException as well?

mrmrinal commented on src/main/java/parser/CommandParser.java:
```diff
@@ -0,0 +1,61 @@
+        String commandKeyword = words[COMMAND_INDEX].toLowerCase();
+
+        switch (commandKeyword) {
+        case FindCommand.KEYWORD:
```
Perhaps you can consider having a default Enum/Case, and have one of the switch statements redirect to it?

mrmrinal commented on src/main/java/parser/TaskParser.java:
```diff
@@ -0,0 +1,96 @@
+     * @param task The task to be parsed.
+     * @return The storage string representation of the task.
+     */
+    public String encode(Task task) {
```
I like your usage of your conditional statements here - very clean and understandable.

mrmrinal commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,97 @@
+            taskList.clear();
+            while (sc.hasNextLine()) {
+                String taskString = sc.nextLine();
+                try {
```
Perhaps you can consider adding multiple catch blocks below the first try block rather than nesting it as the nested catch block is catching a different exception from the main catch block?

mrmrinal commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,109 @@
+        StringBuilder listString = new StringBuilder();
+        int numTasks = tasks.size();
+        for (int i = 0; i < numTasks; i++) {
+            listString.append(i + 1);
```
Perhaps these three statements could be abstracted to a function?

mrmrinal commented on src/test/java/CommandParserTest.java:
```diff
@@ -0,0 +1,90 @@
+
+    @Test
+    public void testDelete() {
+        TaskList t1 = new TaskList();
```
Perhaps the variable names for both tasks could have more appropriate names like ```expectedTaskList``` and ```actualTaskList```?

mrmrinal reviewed (commented):
Hi Jennifer, I have been assigned to to review your PR to checkout the code quality. Overall, I thought your code was very readable and understandable, but I did make a few comments that you can read below!

Hope this helps! 😃

jennibearduit committed:
Refactor JUnit test

jennibearduit committed:
The current code does not have any assertions.

Adding assertions allow developers to test assumptions about the program.

Let's add assertions to some of the existing methods.

jennibearduit committed:
Add a help command.

Current code do not contain a help command, making it hard for
users to know the available commands and their formats.

Adding a help command makes it easier to use the app.

Let's, create a help command to print the formats of all valid
commands.

jennibearduit committed:
Remove several comments to improve code quality.

There were several redundant comments in the command classes.

Removing those comments improve code quality.

jennibearduit committed:
Merge pull request #1 from jennibearduit/branch-A-CodeQuality

Week 5 - Code Quality

jennibearduit committed:
Merge pull request #2 from jennibearduit/branch-BCD-Extension

Week 5 - C Help

jennibearduit committed:
Merge branch 'master' into branch-A-Assertions

jennibearduit committed:
Merge pull request #3 from jennibearduit/branch-A-Assertions

Week 5 - Assertions

jennibearduit committed:
Add ui screenshot

jennibearduit committed:
Add user guide

jennibearduit committed:
Update javadocs

jennibearduit committed:
Edit help command output

jennibearduit committed:
Add fixed duration task

jennibearduit committed:
Merge branch 'branch-B-FixedDurationTasks'

jennibearduit committed:
Add fixed duration task into help command

jennibearduit committed:
Edit javadocs

jennibearduit committed:
Update README.md

jennibearduit committed:
Minor edit
